File: src/order-service/src/Orders.Core/Adapters/DatadogTransactionTracker.cs
// Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
// This product includes software developed at Datadog (https://www.datadoghq.com/).
// Copyright 2026 Datadog, Inc.

using System.Diagnostics;
using System.Globalization;
using System.IO.Compression;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;
using Orders.Core.Telemetry;
using Serilog;

namespace Orders.Core.Adapters;

public class DatadogTransactionTracker : ITransactionTracker
{
    private readonly string _apiKey;
    private readonly string _service;
    private readonly string _environment;
    private readonly Uri _pipelineStatsEndpoint;

    private readonly IHttpClientFactory _httpClientFactory;
    private readonly ILogger<DatadogTransactionTracker> _logger;

    public DatadogTransactionTracker(IHttpClientFactory httpClientFactory,
        IConfiguration configuration,
        ILogger<DatadogTransactionTracker> logger)
    {
        ArgumentNullException.ThrowIfNull(configuration);

        _httpClientFactory = httpClientFactory;
        _logger = logger;
        _apiKey = configuration["DD_API_KEY"] ?? string.Empty;
        _service = configuration["DD_SERVICE"] ?? "print-service";
        _environment = configuration["DD_ENV"] ?? "local";
        var ddSite = configuration["DD_SITE"] ?? "datadoghq.com";
        _pipelineStatsEndpoint = new Uri($"https://trace.agent.{ddSite}/api/v0.1/pipeline_stats");
    }

    public async Task TrackTransactionAsync(string transactionId, string checkpoint)
    {
        if (string.IsNullOrEmpty(_apiKey))
        {
            this._logger.LogInformation("DD_API_KEY is not configured. Skipping transaction tracking for transaction {TransactionId} at checkpoint {Checkpoint}.", transactionId, checkpoint);
            return;
        }

        try
        {
            if (Activity.Current is not null)
            {
                Activity.Current.SetTag("dsm.transaction.id", transactionId);
                Activity.Current.SetTag("dsm.transaction.checkpoint", checkpoint);
            }

            var timestampNanos = (DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() * 1_000_000L)
                .ToString(CultureInfo.InvariantCulture);

            var payload = new
            {
                transactions = new[]
                {
                    new
                    {
                        transaction_id = transactionId,
                        checkpoint,
                        timestamp_nanos = timestampNanos
                    }
                },
                service = _service,
                environment = _environment
            };

            var json = JsonSerializer.Serialize(payload);
            var compressed = Gzip(Encoding.UTF8.GetBytes(json));

            using var client = _httpClientFactory.CreateClient();
            using var request = new HttpRequestMessage(HttpMethod.Post, _pipelineStatsEndpoint);
            request.Headers.Add("DD-API-KEY", _apiKey);
            request.Content = new ByteArrayContent(compressed);
            request.Content.Headers.ContentType = new MediaTypeHeaderValue("application/json");
            request.Content.Headers.ContentEncoding.Add("gzip");

            var response = await client.SendAsync(request);
            response.EnsureSuccessStatusCode();

#pragma warning disable
            _logger.LogInformation(
                "Successfully tracked transaction {TransactionId} at checkpoint {Checkpoint} with status code {StatusCode}",
                transactionId, checkpoint, response.StatusCode);
#pragma warning enable
        }
#pragma warning disable CA1031
        catch (Exception ex)
#pragma warning restore CA1031
        {
            this._logger.LogError(ex, "Failed to track transaction {TransactionId} at checkpoint {Checkpoint}.", transactionId, checkpoint);
        }
    }

    private static byte[] Gzip(byte[] data)
    {
        using var output = new MemoryStream();
        using (var gzip = new GZipStream(output, CompressionMode.Compress))
        {
            gzip.Write(data);
        }

        return output.ToArray();
    }
}

File: src/inventory-service/inventory-core/src/main/java/com/inventory/core/adapters/DatadogTransactionTracker.java
/*
 * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2024 Datadog, Inc.
 */

package com.inventory.core.adapters;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.inventory.core.TransactionTracker;
import com.inventory.core.config.AppConfig;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.ByteArrayOutputStream;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.List;
import java.util.zip.GZIPOutputStream;

@ApplicationScoped
public class DatadogTransactionTracker implements TransactionTracker {

    private static final Logger logger = LoggerFactory.getLogger(DatadogTransactionTracker.class);
    private static final Duration REQUEST_TIMEOUT = Duration.ofSeconds(5);

    private final HttpClient httpClient;
    private final ObjectMapper objectMapper;
    private final AppConfig appConfig;

    @Inject
    public DatadogTransactionTracker(AppConfig appConfig) {
        this.appConfig = appConfig;
        this.objectMapper = new ObjectMapper();
        this.httpClient = HttpClient.newBuilder()
                .connectTimeout(REQUEST_TIMEOUT)
                .build();
    }

    @Override
    public void track(String transactionId, String checkpoint) {
        String apiKey = appConfig.getDdApiKey();
        if (apiKey == null || apiKey.isBlank()) {
            logger.debug("DD_API_KEY not configured, skipping transaction tracking");
            return;
        }

        String url = String.format("https://trace.agent.%s/api/v0.1/pipeline_stats", appConfig.getDdSite());
        long timestampNanos = System.currentTimeMillis() * 1_000_000L;

        var transaction = new Transaction(transactionId, checkpoint, String.valueOf(timestampNanos));
        var payload = new Payload(List.of(transaction), appConfig.getDdService(), appConfig.getEnvironment());

        try {
            byte[] jsonBytes = objectMapper.writeValueAsBytes(payload);
            int statusCode = sendCompressed(jsonBytes, url, apiKey);

            if (statusCode != 202) {
                logger.warn("Datadog pipeline_stats returned unexpected status: {}", statusCode);
            }
        } catch (Exception e) {
            logger.error("Failed to send transaction tracking event for transactionId={}, checkpoint={}", transactionId, checkpoint, e);
        }
    }

    /**
     * Gzip-compresses {@code jsonBytes} and POSTs the result to {@code url}.
     *
     * <p>Required headers:
     * <ul>
     *   <li>{@code Content-Type: application/json}</li>
     *   <li>{@code Content-Encoding: gzip}</li>
     *   <li>{@code DD-API-KEY: <apiKey>}</li>
     * </ul>
     *
     * @return the HTTP response status code
     */
    private int sendCompressed(byte[] jsonBytes, String url, String apiKey) throws Exception {
        var buffer = new ByteArrayOutputStream();
        try (var gzip = new GZIPOutputStream(buffer)) {
            gzip.write(jsonBytes);
        }
        byte[] compressed = buffer.toByteArray();

        var request = HttpRequest.newBuilder(URI.create(url))
                .header("Content-Type", "application/json")
                .header("Content-Encoding", "gzip")
                .header("DD-API-KEY", apiKey)
                .timeout(REQUEST_TIMEOUT)
                .POST(HttpRequest.BodyPublishers.ofByteArray(compressed))
                .build();

        HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
        return response.statusCode();
    }

    // ---------------------------------------------------------------------------
    // Payload records
    // ---------------------------------------------------------------------------

    record Transaction(
            @JsonProperty("transaction_id") String transactionId,
            @JsonProperty("checkpoint")      String checkpoint,
            @JsonProperty("timestamp_nanos") String timestampNanos
    ) {}

    record Payload(
            @JsonProperty("transactions") List<Transaction> transactions,
            @JsonProperty("service")      String service,
            @JsonProperty("environment")  String environment
    ) {}
}
